32-H0ngJu

H0ngJu opened the pull request:
<!-- PR은 최대한 다른 사람이 알아보기 쉽도록 자세히 써주세요. 특히 수도 코드 부분은 더더욱요...!!-->

## 🔗 문제 링크
<!-- 해결한 문제의 링크를 올려주세요. -->
[기타 레슨](https://www.acmicpc.net/problem/2343)

## ✔️ 소요된 시간
<!-- 문제를 해결하는데 소요된 시간을 적어주세요. -->
30분?

## ✨ 수도 코드
<!-- 내가 작성한 코드를 모르는 사람이 봐도 이해할 수 있도록 글로 쉽게 풀어서 설명해주세요. -->

![image](https://github.com/user-attachments/assets/f96fc6bc-a4e6-498c-8b88-72a462b7781e)


이번 문제는 쉽습니다

<br/>
<br/>
<br/>


문제를 읽어보면, '블루레이의 크기에 따라서 수가 정해지겠구나'를 떠올릴 수 있을 것 같습니다.

<br/>
<br/>


블루레이의 수는 1부터 100000까지 가능하고
1부터 100000까지 다 탐색하는 것은 무리라고 생각했습니다.


블루레이가 1개라면 크기는 sum(각 강의의 길이)일거고,
블루레이가 강의의 수만큼 있다면, 크기는 max(강의 길이) 일 것으로 범위를 줄일 수 있습니다.


여기서 탐색 시간을 줄이기 위해서
이분 탐색으로 접근하였습니다.

<br/>
<br/>
<br/>


블루레이 크기 최소 = left = max(lecture)
블루레이 크기 최대 = right = sum(lecture)
브루레이 크기 = mid = (left+right) // 2
로 잡았습니다.


이제 강의들을 하나씩 넣고, mid(블루레이 크기)를 초과하면 레코드 수를 하나씩 늘려갑니다.


최종 레코드 수가 M보다 크면 left를 늘려서 더 많은 강의를 담도록 하고,

반대의 상황이면 rigth를 줄여서 블루레이 범위를 줄입니다!


최종적으로 left를 출력하면 블루레이 크기의 최솟값이 나오게 됩니다


<br/>
<br/>


자료구조였나 알고리즘에서 비슷한 문제를 풀었던 기억이 있어서
쉽게 접근한 것 같네요

<br/>

```python
import sys

def input() : return sys.stdin.readline().rstrip()

N, M = map(int, input().split())
lecture = list(map(int, input().split()))

left = max(lecture)
right = sum(lecture)

while left <= right:
    mid = (left+right)//2
    record_cnt = 1
    time = 0 # 레코드 하나에 들어가는 시간

    for i in range(N):
        if time + lecture[i] > mid: # 시간 초과되면 다음 레코드로
            record_cnt += 1
            time = lecture[i]
        else:
            time += lecture[i]

    if record_cnt <= M: # 깅의가 남아있으면 크기 줄이기
        right = mid - 1
    else:
        left = mid + 1

print(left)
```


## 📚 새롭게 알게된 내용
<!-- 새롭게 알게된 내용이 있다면 작성 해주시고 출처를 남겨주세요. -->

tgyuuAn reviewed (approved):
```python
import sys

def input(): return sys.stdin.readline().rstrip()

N, M = map(int, input().split())
lessons = list(map(int, input().split()))

def check(lessons, mid, M):
    now = 0

    for lesson in lessons:
        if lesson > mid: return False ## <--- 이 부분
        
        if mid >= lesson + now:
            now += lesson
        else:
            now = lesson
            M -= 1
            if M < 0: return False
    return True

left = 0
right = 1_000_000_001
answer = 0
while left+1 < right:
    mid = (left + right) // 2
    if (check(lessons, mid, M-1)):
        answer = mid
        right = mid
    else: 
        left = mid

print(answer)
```

저도 홍쥬르랑 접근은 바로 똭 떠올랐는데,
레슨의 하나의 길이가 블루레이의 크기보다 컸을 때의 경우를 처리해주지 못해서 조금 걸렸습니다.
혹시 홍쥬르의 코드에서는 그게 어떻게 처리되고 있나요 ?

예를 들어 이분탐색에서 블루레이의 크기를 5로 mid값이 나왔을 때
탐색하는 강의의 단일 길이가 10일 경우 어떻게 처리되는 지 궁금해요.

H0ngJu commented:
> ```python
> import sys
> 
> def input(): return sys.stdin.readline().rstrip()
> 
> N, M = map(int, input().split())
> lessons = list(map(int, input().split()))
> 
> def check(lessons, mid, M):
>     now = 0
> 
>     for lesson in lessons:
>         if lesson > mid: return False ## <--- 이 부분
>         
>         if mid >= lesson + now:
>             now += lesson
>         else:
>             now = lesson
>             M -= 1
>             if M < 0: return False
>     return True
> 
> left = 0
> right = 1_000_000_001
> answer = 0
> while left+1 < right:
>     mid = (left + right) // 2
>     if (check(lessons, mid, M-1)):
>         answer = mid
>         right = mid
>     else: 
>         left = mid
> 
> print(answer)
> ```
> 
> 저도 홍쥬르랑 접근은 바로 똭 떠올랐는데, 레슨의 하나의 길이가 블루레이의 크기보다 컸을 때의 경우를 처리해주지 못해서 조금 걸렸습니다. 혹시 홍쥬르의 코드에서는 그게 어떻게 처리되고 있나요 ?
> 
> 예를 들어 이분탐색에서 블루레이의 크기를 5로 mid값이 나왔을 때 탐색하는 강의의 단일 길이가 10일 경우 어떻게 처리되는 지 궁금해요.


다시 문제+코드를 정독하고 왔습니다

태규님 코드는 단일 레코드가 mid보다 크면 check가 False로 반환되어 왼쪽 포인터를 증가 시키는 것 같아요.

제 코드에서는 현재 블루레이크기(mid)보다 단일 강의가 크면,
record_cnt를 늘려줬고, 이 record_cnt와 M을 다시 비교하여 mid(현재 블루레이 크기)를 조절하는 식으로 구현하였습니다. 

```python
for i in range(N):
        if time + lecture[i] > mid: # 시간 초과되면 다음 레코드로
            record_cnt += 1
            time = lecture[i]
```

요 부분이 되겠네용
